Auto registration

nilsdeppe opened the pull request:
## Proposed changes

The main change in this PR is no longer needing to manually register any chare, entry method, or custom reduction function. This is now all done automatically for users.

- Remove unused tests
- Remove remaining Tentacle terminology
- Add auto registration with Charm++ for all chares, entry methods, and custom reduction functions.
- The layer on Charm++ is now documented in the Charm++ Extensions doxygen group.
- Add documentation of the various concepts for the parallelization infrastructure. This is don in response to @nilsleiffischer 's questions. There is still work to be done on this, but I am looking for feedback on what is currently there (except the incomplete Actions discussion).

What breaks:
- All executables using our framework now will need to specify a `charmxx_main_component`
- All chares using our framework need to include `Parallel/CharmRegistration.hpp`
- All chares (not parallel components, those are already handled) using our framework need to auto register, typically done in the destructor. See `Parallel/ConstGlobalCache.hpp` for an example.
- Custom reductions new use a different overload of the `contribute_to_reduction` function.
- Reduction actions must return void
- simple actions must return void unless they take an empty DataBox and return an `initial_databox`.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [x] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...
-->

nilsdeppe committed:
Auto register simple actions with Charm++

This commit adds a factory-style mechanism to automatically register Actions
called via the simple_action method. The registration is done by having
a static member variable of a class generate a pointer array of base-class
pointers that point to derived classes with methods that register the
methods. This means the 'simple_actions' typelist in Parallel
Components is no longer necessary.

nilsdeppe committed:
Improve documentation for writing a new executable.

The documentation describes metavariables, parallel components, phases, the
algorithm, and actions.

nilsdeppe committed:
Remove Test_Main which is not compiled

nilsdeppe committed:
Remove remaining "tentacle" terminology

nilsdeppe committed:
Auto register custom reduction functions with Charm++

Adds support for automatically registering custom reduction functions.

nilsdeppe committed:
Auto register reduction actions with Charm++

nilsdeppe committed:
Auto register parallel components with Charm++

nilsdeppe committed:
Auto register receive data entry method with Charm++

nilsdeppe committed:
Clean up and document CharmMain.cpp

nilsdeppe committed:
Auto register any chare, including main chare

All non-component chares are registered automatically by adding the following
to the destructor:

(void)Parallel::charmxx::RegisterChare<
        ChareName<Ts...>, CkIndex_ChareName<Ts...>>::registrar;

Now when CkNew is called the chare will automatically be registered with
Charm++.

The main chare is special and will be auto registered by specifying the type
alias charmxx_main_component before including CharmMain.cpp. The main component
must have a constructor with the signature:
ConstructorName(const Parallel::charmxx::MainChareRegistrationConstructor)
so it is properly registered.

wthrowe commented:
Per our discussion yesterday, can we have a summary of what this breaks and how to fix it?  Nothing in the "Proposed changes" section is (to me) obviously breaking, and I haven't had a chance to go through the whole PR yet.

kidder commented on src/Parallel/CharmRegistration.hpp:
```diff
@@ -0,0 +1,217 @@
+ *
+ * The reason for using an abstract base class mechanism is that we need to be
+ * able to register entry method templates. The derived classes keep track of
+ * the template parameters and override a the `register_with_charm` function.
```
"a the" -> "the"

kidder commented on src/Parallel/CharmRegistration.hpp:
```diff
@@ -0,0 +1,217 @@
+// clang-tidy: redundant declaration
+template <typename ParallelComponent, typename Action>
+bool Parallel::charmxx::RegisterSimpleAction<ParallelComponent,
+                                             Action>::registerer =  // NOLINT
```
registerer -> registrar

kidder commented on tests/Unit/Parallel/CMakeLists.txt:
```diff
@@ -1,6 +1,90 @@
+  module_Main
+  )
+
+target_link_libraries(
```
can you make a cmake function so it doesn't look so repetitive

kidder commented on tests/Unit/Parallel/CMakeLists.txt:
```diff
@@ -92,11 +176,85 @@ add_dependencies(test-executables Test_ConstGlobalCache)
+add_test(NAME "\"Integration.Parallel.AlgorithmCore\""
+  COMMAND ${CMAKE_BINARY_DIR}/bin/Test_AlgorithmCore)
+
+set_tests_properties(
```
again, cmake function to simplify?

kidder commented on src/Parallel/CharmMain.cpp:
```diff
@@ -171,12 +176,36 @@ void register_parallel_components(tmpl::list<ParallelComponents...> /*meta*/) {
+void register_custom_reducer_functions() noexcept {
+  for (size_t i = 0; i < charm_reducer_functions_size; ++i) {
+    // clang-tidy: do not use pointer arithmetic
+    charm_reducer_functions.insert(
```
try emplace

kidder commented on tests/Unit/Parallel/Test_AlgorithmReduction.cpp:
```diff
@@ -23,7 +23,7 @@ struct TestMetavariables;
 struct SingletonParallelComponent;
 
-static constexpr int number_of_1d_array_elements = 14;
+static constexpr int number_of_1d_array_elements = 46;
```
comment why 46

kidder commented on src/Parallel/Algorithm.hpp:
```diff
@@ -504,6 +504,11 @@ template <typename ParallelComponent, typename ChareType,
 AlgorithmImpl<ParallelComponent, ChareType, Metavariables,
               tmpl::list<ActionsPack...>, ArrayIndex,
               InitialDataBox>::~AlgorithmImpl() {
+  // We place the registerer in the destructor since every AlgorithmImpl will
```
registerer -> registrar in comments too

kidder commented on src/Parallel/CharmRegistration.hpp:
```diff
@@ -82,6 +82,51 @@ struct RegistrationHelper {
 
   virtual void register_with_charm() const noexcept = 0;
   virtual std::string name() const noexcept = 0;
+  virtual bool is_register_chare() const noexcept { return false; };
```
is_registering_chare

kidder commented on src/Evolution/Executables/ScalarWave/CMakeLists.txt:
```diff
@@ -14,8 +14,9 @@ function(add_scalar_wave_executable DIM)
     "// See LICENSE.txt for details.\n"
     "\n"
     "#include \"Evolution/Executables/ScalarWave/EvolveScalarWave.hpp\"\n"
+    "#include \"Parallel/Main.hpp\"\n"
```
"a the"  -> "the" in commit message

kidder commented on tests/Unit/ActionTesting.hpp:
```diff
@@ -67,6 +67,16 @@ class MockProxy {
+    ERROR(
+        "Should not try to serialize the mock proxy. If you encountered this "
+        "error you are using the mocking framework in a way that it was not "
+        "intended to be used. It may be possible to extended to more use cases "
```
extend it

nilsdeppe commented:
Rebased on top of develop

nilsdeppe commented:
doc: https://nilsdeppe.github.io/spectre/pr_578/

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -168,42 +168,231 @@ initialization for the user that can otherwise lead to difficult to debug
-1. `using chare_type` is set to one of `Parallel::Algorithms::Singleton`,
-   `Parallel::Algorithms::Array`, `Parallel::Algorithms::Group`, or
-   `Parallel::Algorithms::Nodegroup`. What these mean is explained below.
+patterns. %Parallel Components are input arguments to the compiler, which then
```
difficult-to-debug

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -168,42 +168,231 @@ initialization for the user that can otherwise lead to difficult to debug
+SpECTRE takes a different approach to input options passed to an executable than
+is common. SpECTRE not only reads an input file at runtime but also has many
+choices made at compile time. The compile time options are specified by what is
+referred as the metavariables. What exactly the metavariables specifies depends
```
referred **to** as the metavariables

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -168,42 +168,231 @@ initialization for the user that can otherwise lead to difficult to debug
+
+### Phases of an Execution
+
+Global synchronization points, where all cores wait for each other are
```
add comma after `other`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -168,42 +168,231 @@ initialization for the user that can otherwise lead to difficult to debug
+
+Another restriction is that currently all Algorithms (described below) must
+execute the same %Actions (also described below) in all phases. This restriction
+is also planned on being relaxed if the need arises.
```
Describe what is special about `Phase::Initialization` and `Phase::Exit`

nilsvu replied:
Yeah it would be good to have a bit of an overview of how the executable is launched in the Charm++ infrastructure, i.e. where the `initialize` function on the parallel components are called.

nilsvu replied:
It is a bit unclear how the algorithm ties phases and parallel components together I think. You could move this sentence about executing the same actions (perhaps substitute parallel components for actions) into the _Algorithm_ section, and provide a few more details.

nilsdeppe replied:
Okay, I'm moving this to the Algorithm section at the end as a note. I think it'll be clearer what this means after the algorithm is explained, and moving it into a note should indicate that it's not essential to understand that. I added this because Mark had asked about it at some point and want to make sure it's documented so if someone else wants this in the future they at least have some starting point.

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -168,42 +168,231 @@ initialization for the user that can otherwise lead to difficult to debug
+
+Since most numerical algorithms repeat steps until some criterion such as the
+final time, or converge is met, SpECTRE's parallel components are designed to do
+such iterations for the user. An Algorithm executes an order list of actions
```
order -> ordered

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -168,42 +168,231 @@ initialization for the user that can otherwise lead to difficult to debug
+### The Algorithm
+
+Since most numerical algorithms repeat steps until some criterion such as the
+final time, or converge is met, SpECTRE's parallel components are designed to do
```
converge -> convergence
Remove comma after `time`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -168,42 +168,231 @@ initialization for the user that can otherwise lead to difficult to debug
+1. `using chare_type` is set to one of:
+   1. `Parallel::Algorithms::Singleton`s have one object in the entire execution
+      of the program.
+   2. `Parallel::Algorithms::Array`s hold zero or more elements each of which is
```
comma after `each`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -168,42 +168,231 @@ initialization for the user that can otherwise lead to difficult to debug
+   global metavariables struct. Examples of this technique are given below.
+3. `using action_list` is set to a `tmpl::list` of the %Actions (described
+   below) that the Algorithm running on the %Parallel Component executes. The
+   %Actions are executed in the order that they are in in the `tmpl::list`.
```
in in -> given in

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -168,42 +168,231 @@ initialization for the user that can otherwise lead to difficult to debug
-   written a general implementation that provides this functionality; all that
-   you need to provide is a plain-old-data struct of the size of at most 3
-   integers.
+   simple action called during the `initialization` Phase.
```
`Initialization`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -168,42 +168,231 @@ initialization for the user that can otherwise lead to difficult to debug
+`Parallel::ArrayIndex1D` through `Parallel::ArrayIndex6D`. When writing custom
+array indices, the Charm++ manual tells you to write your own `CkArrayIndex`,
+but we have written a general implementation that provides this functionality;
+all that you need to provide is a plain-old-data struct of the size of at most 3
```
after `plain-old-data` add ([POD](http://en.cppreference.com/w/cpp/concept/PODType))

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -168,42 +168,231 @@ initialization for the user that can otherwise lead to difficult to debug
+phase, after which the `execute_next_phase` function gets called. The
+`execute_next_phase` function determines what the %Parallel Component should do
+during the next phase. For example, it may simply call `perform_algorithm`, call
+a series of simple Actions, perform a reduction over an Array, or not do
```
`Actions` -> `actions`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -168,42 +168,231 @@ initialization for the user that can otherwise lead to difficult to debug
+
+An example of an array %Parallel Component is:
+\snippet Test_AlgorithmParallel.cpp array_parallel_component
+Elements are inserted into the Array by using the Charm++ `insert` member
```
link to charm++ manual http://charm.cs.illinois.edu/help

nilsdeppe committed:
Simple and reduction actions return void or InitialDataBox

If simple actions could return any DataBox then the algorithm would become
chaotic to keep right since simple actions can be invoked at any time. Thus,
simple actions are used only for initializing the DataBox and for mutating
values in the DataBox.

nilsdeppe committed:
Rename explicit_single_action to simple_action

nilsdeppe committed:
Only update Algorithm headers when necessary

Previously the algorithm ci files and headers, and charm++ generated headers
were "updated" every time CMake was run. This resulted in unnecessary rebuilds
of executables. We now only rebuild if the files on disk are not up-to-date with
what would be written to disk.

nilsdeppe committed:
Use configure_file for ScalarWave executable

Previously the scalar wave executable file was rewritten every time CMake was
ran in such a way that triggered a rebuild. The executables are now only rebuilt
if the contents of the files change.

nilsdeppe commented:
I've added 4 commits:
- Require simple and reduction actions to return void. Simple actions that take an empty DataBox may return the `initial_databox`
- Return `explicit_single_action` to `simple_action`
- Stop updating generated algorithm ci and hpp files when it's not necessary. This caused unnecessary rebuilds after running CMake
- Do a similar thing with the EvolveScalarWave executables so they aren't rebuilt as often.

I've addressed the comments from the last time we looked at the documentation and also have the Actions documentation ready to be looked at: http://gh.nilsdeppe.com/spectre/pr_578/group__ParallelGroup.html @nilsleiffischer can you please look at the documentation and give feedback? It should answer a lot of the questions you've had so far.

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
-invocation of actions. This approach allows us to eliminate the need for users
-to work with Charm++'s interface files, which are very error prone and difficult
-to use.
+The return type is discussed at the end of the %Actions section. Simple actions
```
at the end of each action type's section.

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
-#### 1. Simple %Actions
+\snippet Test_AlgorithmNestedApply1.cpp bad_recursive_call
+
+`ckLocal()`  is a Charm++ provided method that returns a pointer to the local
```
"Here `ckLocal`"

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+retrieved from the ConstGlobalCache. The remote entry method invocations are
+slightly different for different types of actions, so they will be discussed
+below. However, one thing that is disallowed for all actions is calling an
+action from within an action on the same parallel component that is not invoked
```
"action locally from within an action on the same parallel component. Specifically,"

Add a few sentences after the example describing how going through the RTS is fine

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+(currently executing) parallel component. See the
+[Charm++ manual](http://charm.cs.illinois.edu/help) for more information.
+
+Array parallel components can have actions invoked in two ways. First is a
```
"Array, group, and nodegroup"

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+array element's index. The below example shows how a broadcast would be done
+manually by looping over all elements in the array:
+
+\snippet Test_AlgorithmParallel.cpp call_on_indexed_array
```
Move comment block in example out of example

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
-`is_ready` function is:
+functions of classes. They are the direct analog of entry methods in Charm++
+except that the member data is stored in the `db::DataBox` that is passed in as
+the first argument. There are a couple important things to note with simple
```
couple *of* important

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+the first argument. There are a couple important things to note with simple
+actions:
+
+1. A simple action's must return void but can use `db::mutate` to change values
```
remove `'s`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+
+1. A simple action's must return void but can use `db::mutate` to change values
+   of items in the DataBox if the DataBox is taken as a non-const reference.
+   There is one exception: if the input DataBox is empty, then the explicit
```
explicit singe -> simple

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+   of items in the DataBox if the DataBox is taken as a non-const reference.
+   There is one exception: if the input DataBox is empty, then the explicit
+   single action can return a DataBox of type `initial_databox`. That is, an
+   action taking an empty DataBox and return the `initial_databox` is
```
return -> returning

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+   effectively constructing the DataBox in its initial state.
+2. A simple action is instantiated once for an empty
+   `db::DataBox<tmpl::list<>>`, once for a DataBox of type
+   `initial_databox_type` (listed in the parallel component), and once for each
```
remove `_type`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+   returned DataBox from the iterable actions in the `action_list` in the
+   parallel component. In some cases you will need specific items to be in the
+   DataBox otherwise the action won't compile. To restrict which DataBoxes can
+   be passed you should user `Requires` in the action's `apply` function
```
user -> use

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+   DataBox otherwise the action won't compile. To restrict which DataBoxes can
+   be passed you should user `Requires` in the action's `apply` function
+   template parameter list. For example,
+   \snippet Test_AlgorithmCore.cpp requires_action
```
Add a `where` explaining the conditional

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+   template parameter list. For example,
+   \snippet Test_AlgorithmCore.cpp requires_action
+
+A simple action that does not take any arguments can be called using a `CProxy`
```
Wat indentation

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+\note
+You must be careful about type deduction when using `std::make_tuple` because
+`std::make_tuple(0)` will be of type `std::tuple<int>`, which will not work if
+the action is expecting to receive a `size_t` as its extra argument.
```
"Instead, you can do it in one of two ways."

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
-operator (`[array_index]`).
+The `inboxes` is a collection of the tags passed to `receive_data` and are
+specified in the iterable actions member type alias `inbox_tags`, which must be
+a `tmpl::list`. The `inbox_tags` must have two member type alias, a
```
Plural: type alias -> type aliases

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+a `tmpl::list`. The `inbox_tags` must have two member type alias, a
+`temporal_id` which is used to identify when the data was sent, and a `type`
+which is the data to be stored in the `inboxes`. The types are typically a
+`std::unordered_map<temporial_id, DATA>`. In the discussed scenario of waiting
```
`temporal_id`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+`std::unordered_map<temporial_id, DATA>`. In the discussed scenario of waiting
+for neighboring elements to send their data the `DATA` type would be a
+`std::unordered_map<TheElementIndex, DataSent>`. Having `DATA` be a
+`std::undordered_multiset` is currently also supported. Here is an example of a
```
fix`und`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+specified in the iterable actions member type alias `inbox_tags`, which must be
+a `tmpl::list`. The `inbox_tags` must have two member type alias, a
+`temporal_id` which is used to identify when the data was sent, and a `type`
+which is the data to be stored in the `inboxes`. The types are typically a
```
which is the **type of** data

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+longer be executed, and choose which action in the `ActionList`/`action_list` to
+execute next. This is all done via the return value from the `apply` function.
+The `apply` function for iterable actions must return a `std::tuple` of one,
+two, or three elements. The first element of the tuple is the new DataBox type,
```
remove `type`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+unnecessary copying of the DataBox. The second argument is an optional bool, and
+controls whether or not the algorithm is terminated. If the bool is `true` then
+the algorithm is terminated, by default it is `false`. Here is an example of how
+to return a DataBox with the same type that is passed in and also terminated
```
terminated -> terminate

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+`perform_algorithm` or `receive_data` methods.
+
+The third optional element in the returned `std::tuple` is a `size_t` whose
+integer value corresponds to the index of the action to be called next in the
```
remove `integer`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+The third optional element in the returned `std::tuple` is a `size_t` whose
+integer value corresponds to the index of the action to be called next in the
+`action_list`. The metafunction `tmpl::index_of<list, element>` can be used to
+get an `integral_constant` with the value of the index of the element `element`
```
`tmpl::integral_constant`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+
+Iterable actions are invoked as part of the algorithm and so the only way
+to request they be invoked is by having the algorithm run on the parallel
+component. The algorithm can be explicitly triggered by call the
```
triggered -> evaluated

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+
+\snippet Test_AlgorithmReduction.cpp contribute_to_reduction_example
+
+in the array. This reduces over the parallel component
```
remove `in the array`?

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+in the array. This reduces over the parallel component
+`ArrayParallelComponent<Metavariables>`, reduces to the parallel component
+`SingletonParallelComponent<Metavariables>`, and calls the action
+`singleton_reduce_sum_int`. The `CkReduction::sum_int` argument tells Charm++
```
`after the reduction has been perform`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+that you want to sum integers instead of some other type or operation. The
+reduction action is:
+
+\snippet Test_AlgorithmReduction.cpp singleton_reduce_sum_int
```
make `value` const

nilsvu replied:
The prefix `singleton` doesn't make it clear that this is an action. Just remove the prefix, or put the keyword `action` somewhere in the function name. Also the name suggests that it does the reduction, but it is only the receiver. And it is a type, so it should be camel case. So perhaps better call it `ProcessReducedSum` or similar.
Edit: To be clear, I am referring to the snippet _content_, not its name.

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+\snippet Test_AlgorithmReduction.cpp contribute_to_broadcast_reduction
+
+It is often necessary to reduce custom data types, such as `std::vector` or
+`std::unordered_map`. Charm++ support such custom reductions, and so does our
```
plural support -> supports

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+
+The action that is invoked with the result of the reduction is:
+
+\snippet Test_AlgorithmReduction.cpp custom_reduction_action
```
const the value

nilsvu commented on docs/GroupDefs.hpp:
```diff
@@ -159,155 +164,485 @@
+SpECTRE takes a different approach to input options passed to an executable than
+is common. SpECTRE not only reads an input file at runtime but also has many
+choices made at compile time. The compile time options are specified by what is
+referred to as the metavariables. What exactly the metavariables specifies
```
_the metavariables specify_ or _the metavariables struct specifies_

nilsdeppe replied:
Going with the second since metavariables is singular so the first is not correct :)

nilsvu commented on docs/GroupDefs.hpp:
```diff
@@ -159,155 +164,485 @@
+  that are to be created. Most evolution executables will have the
+  `DgElementArray` parallel component listed. An example of a `component_list`
+  for one of the test executables is:
+  \snippet Test_AlgorithmCore.cpp component_list_example
```
Perhaps show a snippet of `EvolveScalarWave` instead, that is closer to common use cases?

nilsdeppe replied:
We don't allow snippets from inside the production code currently. The reason being that production code may not be compiled, whereas the tests will always be compiled and tested. I could suggest people look at certain files in production code, but then everything there should be explained here. Maybe we should tell people to look at a production example that we document heavily inside with C++ comments?

I'm happy to add more of an explanation, what do you think should be covered?

nilsvu replied:
Yeah let’s do that, let’s pick a particular example (`EvolveScalarWave`) and document it heavily.

Instead of the snippet you could briefly describe a common use case, e.g. have one parallel component perform an actions sequence on an array of elements, and another to write data to disk. This gives a more intuitive picture of what these structures are intended for.

nilsvu commented on docs/GroupDefs.hpp:
```diff
@@ -159,155 +164,485 @@
+receiving data from various components to allow for more complex decision
+making. Here is an example of a `determine_next_phase` function and the `Phase`
+enum class:
+\snippet Test_AlgorithmCore.cpp determine_next_phase_example
```
Perhaps also show a snippet from `EvolveScalarWave` instead.

nilsdeppe replied:
Same response about no snippets in production code

nilsvu replied:
Then I suggest to also add a brief note on a common use case, e.g. mention phases `FindInitialData` and `Evolve`.

nilsvu commented on docs/GroupDefs.hpp:
```diff
@@ -159,155 +164,485 @@
+patterns. %Parallel Components are input arguments to the compiler, which then
+writes the parallelization infrastructure that you requested for the executable.
+There is no restriction on the number of %Parallel Components, though
+practically it is best to have around 10 at most.
```
It would be good to give a short overview here of the concepts to be explained in detail below. I suggest a list like:

- Metavariables: Provides high-level configuration to the compiler, e.g. the physical system to be simulated.
- Phase: Defines distinct simulation phases separated by a global synchronisation point, e.g. `Initialize`, `Evolve` and `Exit`.
- Algorithm: In each phase, asks all parallel components to execute their actions.
- Parallel component: Maintains and executes its actions, e.g. repeats a sequence of actions until a condition is met.
- Action: Performs a small computational task, e.g. computing an integral over a small domain. May require input from other actions.

nilsvu commented on docs/GroupDefs.hpp:
```diff
@@ -159,155 +164,485 @@
+
+### %Actions
+
+For those familiar with Charm++, actions should be thought of as effectively
```
Perhaps better start with a more approachable description of actions, instead of referring to the Charm++ concepts, since this will probably not help the reader at this point.

nilsdeppe replied:
The sentence is meant to allow anyone who knows about Charm++ to be able to get up to speed quicker. I could add something to the end of the sentence like "if not don't worry."

nilsvu replied:
My suggestion is to start more approachable, then give the reference to charm++ to help readers familiar with that.

nilsvu commented on docs/GroupDefs.hpp:
```diff
@@ -159,155 +164,485 @@
+`Parallel::ReductionData`, which takes as template parameters the types it
+holds, similar to `std::tuple`. An example of a custom reduction function is:
+
+\snippet Test_AlgorithmReduction.cpp custom_reduce_function
```
Perhaps `reduce_custom_data` would be a better name.

nilsdeppe replied:
This comment has nothing to do with the documentation. The snippet name is largely irrelevant, and the current one I think is actually clearer than what you suggested. The rendered function has the name `reduce_reduction_data`.

nilsvu replied:
_Nothing to do with documentation_ is a bit of an overstatement, it’s rendered in the documentation after all, and tests serve also an illustrative purpose. Then take this comment, and the previous, as a suggestion to adjust the Test code.

nilsdeppe replied:
Not really, the function name you are talking about already has the name `reduce_reduction_data`, are you asking that be changed? I'm not changing the snippet name.

wthrowe replied:
I'm confused by this discussion.  Are you two sure you're talking about the same thing?

nilsvu replied:
I see the confusion now. Agreed, the snippet name is irrelevant :D I simply suggested to make this snippet‘s content, which is pulled from a test, more expressive. In fact, if you prefer the naming of this particular function to remain as it is I would not object, but I think the snippet `singleton_reduce_sum_int` could be improved, as I commented above.

nilsdeppe replied:
Okay, so at least we agree on that :) I've changed the `single_reduce_sum_int` (not pushed yet). I'm fine changing the name of the function as well, was just confused.

nilsdeppe replied:
After thinking about it more, I like `reduce_reduction_data` since the type it's reducing over is a `ReductionData`

nilsvu reviewed (changes requested):
It's great to have this extended documentation on the parallelisation architecture. I suggest you add a link under `Tutorials`.

nilsdeppe commented:
I also updated the rendered documentation and rebased on develop
